internal/services/admin/v1/server.go

```diff
@@ -234,49 +240,100 @@ func (a *AdminServiceImpl) ReplayTasks(ctx context.Context, req *contracts.Repla
+		record := v1.TaskIdInsertedAtRetryCount{
 			Id:         task.ID,
 			InsertedAt: task.InsertedAt,
 			RetryCount: task.RetryCount,
+		}
+		tasksToReplay = append(tasksToReplay, TaskIdInsertedAtRetryCountWithWorkflowRunId{
+			TaskIdInsertedAtRetryCount: record,
+			WorkflowRunId:              task.WorkflowRunExternalID,
 		})
-	}
 
-	// FIXME: group tasks by their workflow run id, and send in batches of 50 workflow run ids...
+		taskIdInsertedAtRetryCountToExternalId[record] = task.WorkflowRunExternalID
+	}
 
-	// send the payload to the tasks controller, and send the list of tasks back to the client
-	toReplay := tasktypes.ReplayTasksPayload{
-		Tasks: tasksToReplay,
+	workflowRunIdToTasksToReplay := make(map[pgtype.UUID][]v1.TaskIdInsertedAtRetryCount)
+	for _, item := range tasksToReplay {
+		workflowRunIdToTasksToReplay[item.WorkflowRunId] = append(
+			workflowRunIdToTasksToReplay[item.WorkflowRunId],
+			item.TaskIdInsertedAtRetryCount,
+		)
 	}
 
-	msg, err := msgqueue.NewTenantMessage(
-		sqlchelpers.UUIDToStr(tenant.ID),
-		"replay-tasks",
-		false,
-		true,
-		toReplay,
-	)
+	var batches [][]v1.TaskIdInsertedAtRetryCount
+	var currentBatch []v1.TaskIdInsertedAtRetryCount
+	batchSize := 100
 
-	if err != nil {
-		return nil, err
+	for _, tasksForWorkflowRun := range workflowRunIdToTasksToReplay {
+		if len(currentBatch) > 0 && len(currentBatch)+len(tasksForWorkflowRun) > batchSize {
+			// If the current batch would exceed the batch size if we added the current workflow run's tasks,
+			// we "finalize" the batch and start a new one
+			batches = append(batches, currentBatch)
+			currentBatch = nil
+		}
+
+		if len(tasksForWorkflowRun) > batchSize {
+			// If the current workflow run's task count exceeds the batch size on its own,
+			// we let it be its own batch
+			batches = append(batches, tasksForWorkflowRun)
+		} else {
+			// Otherwise, add it to the current batch
+			currentBatch = append(currentBatch, tasksForWorkflowRun...)
+		}
 	}
 
-	err = a.mq.SendMessage(ctx, msgqueue.TASK_PROCESSING_QUEUE, msg)
+	if len(currentBatch) > 0 {
+		// Last case to handle - add the last batch if it has any tasks
+		batches = append(batches, currentBatch)
+	}
 
-	if err != nil {
-		return nil, err
+	replayedIds := make([]string, 0)
+
+	for _, batch := range batches {
+		// send the payload to the tasks controller, and send the list of tasks back to the client
+		toReplay := tasktypes.ReplayTasksPayload{
+			Tasks: batch,
+		}
+
+		msg, err := msgqueue.NewTenantMessage(
+			sqlchelpers.UUIDToStr(tenant.ID),
+			"replay-tasks",
+			false,
+			true,
+			toReplay,
+		)
+
+		if err != nil {
+			return nil, err
+		}
+
+		err = a.mq.SendMessage(ctx, msgqueue.TASK_PROCESSING_QUEUE, msg)
+
+		if err != nil {
+			return nil, err
+		}
+
+		for _, task := range batch {
+			replayedIds = append(replayedIds, taskIdInsertedAtRetryCountToExternalId[task].String())
+		}
+
+		time.Sleep(200 * time.Millisecond)
```

Comment:
do we need to sleep here? and for how long if we do?

Reply:
this seems reasonable